saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/node/Node.scala

```diff
@@ -85,42 +87,47 @@ class Node[T <: AnyRef](val keyFunc: T => Any = (x: T) => x, val tag: ClassTag[T
     for (e <- outgoing) e.clear
   }
 
-  private var count = 0
-
-  private def incrementCount(): Int = this.synchronized {
-    val ret = count
-    count = count + 1
-    ret
-  }
-
-  private def decreaseCount(): Int = this.synchronized {
-    val ret = count
-    count = count - 1
-    ret
-  }
+  private var count: AtomicInteger = new AtomicInteger()
 
   def contains(t: T): Boolean = collection.contains(toNT(t))
 
   def toNT(t: T): NT = new NT(t, keyFunc)
 
   private def containsNT(nt: NT): Boolean = collection.contains(nt)
 
-  def addInstance(t: T, train: Boolean = true, populateEdge: Boolean = true) = {
-    val nt = toNT(t)
-    if (!containsNT(nt)) {
-      val order = incrementCount()
-      if (train) this.trainingSet += nt else this.testingSet += nt
-      this.collection += nt
-      this.orderingMap += (order -> nt)
-      this.reverseOrderingMap += (nt -> order)
-      if (populateEdge) {
-        outgoing.foreach(_.populateUsingFrom(t, train))
-        incoming.foreach(_.populateUsingTo(t, train))
-      }
-      joinNodes.foreach(_.addFromChild(this, t, train, populateEdge))
+  /** Adds an instance to the the [[Node]].
+    *
+    * @param instance Node instance.
+    * @param train If the instance is a training instance.
+    * @param populateEdge If populating edges from the current Node.
+    */
+  def addInstance(instance: T, train: Boolean = true, populateEdge: Boolean = true): Unit = {
+    val nodeInstance = toNT(instance)
+
+    if (containsNT(nodeInstance)) {
+      //       logger.warn(s"The instance $t is duplicate and it will be ignored! " +
+      //         s"This might be because you add the same instance to both train and test set. ")
+      return
```

Comment:
Removed the `return` statement and make it `logger.trace`. I do not mind having an `else` statement here as the lines are not too long to cause indent issues.